Openstack provider

zreigz opened the pull request:
Our pull request delivers solution to create Kubernetes cluster on the top of OpenStack. Heat OpenStack Orchestration engine describes the infrastructure for Kubernetes cluster. CentoOS images are used for Kubernetes host machines.

We tested our solution with DevStack and Citycloud provider.

We believe that our solution will fill the gap that which is on the market.

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21737)

<!-- Reviewable:end -->

ohtake committed:
Add Heat template to create a Kubernetes stack on OpenStack

Commit:
OpenStack provider

- Create Kubernetes cluster on the top of OpenStack

zreigz commented:
@ohtake is also a author of those commits

davidopp commented:
ok to test

davidopp commented:
@ohtake needs to sign the CLA.

davidopp commented:
cc/ @ashw7n

ohtake commented:
@davidopp I have just singned the CLA. Could you re-verify the CLA?

bgrant0607 commented:
I don't think the CLA bot ever works for PRs with multiple authors.

bgrant0607 commented:
We're still trying to get 1.2 out the door and have a couple other new cloudproviders to review, so this will need to wait for a little while. 

It would also be useful to have some people using K8s on Openstack to review. I suggest posting to kubernetes-dev to plead for reviewers.

zreigz commented:
I've created post on kubernetes-der regarding to openstack-provider:
https://groups.google.com/forum/#!topic/kubernetes-dev/t1-0zG3Lmeg

I hope someone will be interested.

zreigz commented:
Ok thanks

bgrant0607 commented:
Would it be possible to split this into 2 PRs, each submitted by its author?

idvoretskyi commented:
I suppose the target directory has to be renamed (or pushed into the new subdirectory of "openstack" directory).

The described method is not the only way to deploy Kubernetes cluster on OpenStack (don't forget about Murano, Magnum as well as about some minor projects), therefore it may confuse potential end-users.

ohtake commented:
@bgrant0607 Okay, I will create one after we pick a directory name.

@idvoretskyi Good point. How about `openstack-heat` and `openstack/heat`? I prefer the former for simple hierarchy. What do you think?

ohtake committed:
Use host-gw flannel for better performance

idvoretskyi commented:
@ohtake yes, your proposal if useful - let's do it.

ohtake committed:
Execute salt-call twice

davidopp commented:
cc/ @mikedanese

Commit:
certificate should be created for the virtual kubernetes ip

elsonrodriguez commented:
Tried running through this provisioner, ended up not getting past this portion of the script:
```
heat --api-timeout 60 stack-create -P external_network=public -P ssh_key_name=kubernetes_keypair -P server_image=30ecf289-c028-4a7c-b782-12fe3e20c95b -P master_flavor=m1.small -P minion_flavor=m1.small -P number_of_minions=1 -P max_number_of_minions=1 -P dns_nameserver=8.8.8.8 -P kubernetes_salt_url=http://.../kubernetes/kubernetes-salt.tar.gz -P kubernetes_server_url=http://.../kubernetes/kubernetes-server.tar.gz --template-file kubecluster.yaml KubernetesStack
b'ERROR: Property error: : resources.provision_network_master.properties.config: : Value must be a string'
```
This is with python-heatclient 1.1.0, 1.0.0, and Python 3.4.4 running from OSX.

ohtake commented:
@elsonrodriguez
Thank you for trying the provider.

I could not reproduce it with CentOS 7 64bit, Python **2.7.5**, python-heatclient 1.1.0,
but I could reproduce it with CentOS 7 64bit, Python **3.4.3**, python-heatclient 1.1.0.

I think there is an issue in python-heatclient.
<https://bugs.launchpad.net/python-heatclient/+bug/1498037>
When I execute `heat -d stack-create ...` under Python 3, I got integer arrays.

elsonrodriguez commented:
@ohtake Thanks, python version change fixed that issue. Will keep testing provider, but before I forget I had to change this line (wasn't sure of original intent, guessed an operator).
```
--- a/cluster/openstack/util.sh
+++ b/cluster/openstack/util.sh
@@ -91,7 +91,7 @@ function create-stack() {
   echo "[INFO] Execute commands to create Kubernetes cluster"
   # It is required for some cloud provider like CityCloud where swift client has different credentials
   source "${ROOT}/openrc-swift.sh"
-  if [[ ! -v OS_PROJECT_ID ]]; then
+  if [[ ! -z OS_PROJECT_ID ]]; then
     SWIFT_PROJECT_ID="${OS_TENANT_ID}"
   else
     SWIFT_PROJECT_ID="${OS_PROJECT_ID}"
```

Otherwise I would get a bash error:

```
./../cluster/../cluster/openstack/util.sh: line 94: conditional binary operator expected
```

This is with bash 3.2.57 on OSX.

ohtake commented:
@elsonrodriguez
Reproduced with bash 4.1.2 on CentOS 6.7.
`-v` operator was [added in bash 4.2](http://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash#comment31736775_17538964).

`! -z OS_PROJECT_ID` is always false because argument of `-z` is not varname but string.
I will modify the line `-z "${OS_PROJECT_ID}"`.

cc: @zreigz

ohtake committed:
Make shell script compatible with older bash

zreigz commented:
LGTM Thank for reporting this issue

elsonrodriguez commented:
@zreigz @ohtake Got it working! 

It looks like the cloud provider isn't configured though. I'm going to look through the salt configuration and see if it's feasible. (a working configuration will look sort of like this: https://gist.github.com/elsonrodriguez/8d724baf910a5ad10a7e3b569b0f0b18)

I'm also behind a proxy, so I've had to make some hacks that I'd like to normalize and contribute.

How should we coordinate?

elsonrodriguez committed:
Adding support for provisioning behind a Proxy.

This assumes you have your environement variables set correctly.

When ENABLE_PROXY is set to true, it takes the current proxy
settings and applies them to the heat configuration.

Also modified the defaults system in config-default.sh

elsonrodriguez commented:
@zreigz @ohtake I've gotten the initial support for proxies done: https://github.com/elsonrodriguez/kubernetes/commit/07e78fe2ca52a4d12ce47ea66105ba4684f2aa1d

I need to polish up a few things and add comments, and then I can move on to the cloudprovider.

I'll provide updates regarding the cloudprovider as I make progress.

zreigz commented:
Thanks for your support

elsonrodriguez committed:
Fixed proxy settings for minion, added comments.

elsonrodriguez committed:
Made it so that existing Openstack environment variables are used.

If someone has an openrc as part of their profile, this will make kube-up work automatically.

The only things that have to be modified are in config-default.sh, either by editing the file or setting environment variables.

elsonrodriguez committed:
Adding cloudprovider support for Openstack kube-up provider.

Making the assumption that the person running kube-up has their
Openstack environment setup, those same variables are being passed
into heat, and then into openstack.conf.

The salt codebase was modified to add openstack as well.

elsonrodriguez commented:
@zreigz @ohtake The proxy branch is pretty much wrapped up, I need to do some more testing though. Also I'd like someone to test and see if it still works good on non-proxied setups: https://github.com/elsonrodriguez/kubernetes/commits/openstack-provider-proxy

I've also started on enabling cloudprovider support for Openstack. It passed a quick smoke-test(provisioning a cinder volume): https://github.com/elsonrodriguez/kubernetes/commit/0bf8b946022dfc0b46db5820bd87ee761ecf453e

Take a look and let me know what you think.

elsonrodriguez commented:
@zreigz @ohtake 
I've made good progress on the cloud-provider integration with OpenStack: https://github.com/elsonrodriguez/kubernetes/commit/b4d36474292a86cf6fda7eb7ccda7d5579e0d05b

Also, here is an early preview of the documentation adjustments to reflect the changes I'm proposing, let me know if you like the direction: https://github.com/elsonrodriguez/kubernetes.github.io/commit/852b252361a02dcfc0aa39bd2a747d858c56948b

Give the changes a shot and let me know how it looks, or if you have any questions.

elsonrodriguez committed:
Adding automatic OS image download for Openstack provider.

This makes it so that we download the OS image automatically.

Also contains other usability improvements:
- kubectl context created with heat stack name
- Bumped default minions to 3

elsonrodriguez commented:
@ohtake @zreigz I've sent a Pull Request for my proposed changes: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/8

I think it's ready for testing and merging into this Pull Request so we can move forward and get more feedback.

I ended up having to modify the Salt code, so once my changes are merged in, a rebase will probably need to happen.

elsonrodriguez commented:
@bgrant0607 I'm piling on contributions that affect commits from both @ohtake and @zreigz. How much worse have I made the CLA bot situation? How should I resolve it?

elsonrodriguez committed:
Adding missing variable defaults for OpenStack provider.

elsonrodriguez committed:
Adding openstack example to get-kube.sh

elsonrodriguez commented:
Related Documentation PR here: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes.github.io/pull/1/files?diff=split

elsonrodriguez commented:
@ohtake @zreigz @cheld This PR was discussed during the Openstack Sig meeting: https://docs.google.com/document/d/1iAQ3LSF_Ky6uZdFtEZPD_8i6HXeFxIeW4XtGcUJtPyU/edit?ts=57150425#heading=h.1n8ybxoas1li (search for 'April 19th')

The main takeaways for this PR were that heat is a barrier to entry, and that a more monolithic CLI should be used.

So two main questions: 

- How feasible is it to do this without Heat?
- What's the most commonly used monolithic CLI we can use? (in please of nova, swift, glance)

cheld commented:
@elsonrodriguez. Thanks for the information. Occasionally, I participate in this meeting. But, it is at midnight hours at local time :) and the agenda did not seem relevant to me...

Why is heat a problem? If we do it on our own we start to reimplement heat, don't we? And the provisioning process is run on client side. 

We could think about dropping the creation of the network and leave this for the end-user, and just create the master and the minions....

xsgordon commented:
@cheld so at the core of the discussion was the goal of providing something in-tree in Kubernetes that is as basic a way as possible for someone from the Kubernetes community to get a K8S cluster up and running on an existing OpenStack cloud which they may or may not administer. If they don't and it doesn't have Heat then they are out of luck using this implementation. The feeling was that ideally the in-tree kube-up implementation for OpenStack should be as easy to use and widely deployable as possible.

For folks with more production-like requirements and/or broader flexibility as to which OpenStack projects to install the proposal was that we document the decision tree to help them navigate the other approaches (Heat, Murano, Magnum, etc.). The question posed on Slack was whether modifying these templates to use the basic client calls instead of doing them via Heat would be re-implementing Heat which is a fair question but it seems like only very basic orchestration is being performed here?. This may be more of an issue if the intent is to target a more production-like use case then the "just get me a cluster" user kicking the tires though there are Heat templates elsewhere [2][3][4] that more directly address this?

[1] https://www.openstack.org/assets/survey/April-2016-User-Survey-Report.pdf
[2] https://github.com/larsks/heat-kubernetes
[3] http://git.openstack.org/cgit/stackforge/heat-coe-templates/
[4] http://git.openstack.org/cgit/openstack/magnum/tree/magnum/templates/kubernetes

elsonrodriguez committed:
Improved user experience for users of OpenStack provider behind a proxy.

cheld commented:
@xsgordon So the argument against heat is that not every cloud provider supports it yet, right?

I also think that this kube-up script based approach is rather for Kubernetes developers or people that compile Kubernetes themselves rather than production. 

Currently, we do not have the time to replace heat. What about restructuring the code so that heat could potentially be replaced?

elsonrodriguez commented:
@cheld  If there is no time to replace heat, it's probably best to go forward as-is (no restructuring). Whoever performs the heat-removal work would be better off working from an unmolested base.

elsonrodriguez commented:
@batikanu It may be the case that if you sign the CLA, the bot will be happy. Please sign the [Individual](http://code.google.com/legal/individual-cla-v1.0.html) or [Corporate](http://code.google.com/legal/corporate-cla-v1.0.html) CLA as soon as possible.

cc @cheld @ohtake @zreigz

elsonrodriguez committed:
Enabled Addons for Openstack kube-up provider.

Hard-coded the enabling of the common addons:

- logging
- kube-dashboard
- monitoring

Will make it configureable in a subsequent PR.
Also need to enable configuration of basic_auth.csv

eparis commented:
to solve the CLA problem. Get everyone who wrote a commit to state, here in this PR, they are ok with the commit being merged into kubernetes. And admin will then add the label `cla: human-approved`. With that label you can ignore the `cla: no`

elsonrodriguez committed:
Increasing default node size for kube-up Openstack provider.

The default pod sizes are too small for some Java apps. Also the logging/monitoring pods were taking up too many resources.

batikanu commented:
I have signed the CLA already. Can you check again?

zreigz commented:
After merge and rebase cluster is created but k8s doesn't work because of docker:
```
kwi 22 11:26:36 kubernetesstack-master kube-master-addons.sh[10621]: Failed to restart docker.service: Unit docker.service failed to load: No such file or directory.
```

Commit:
Add configuration for e2e tests

quinton-hoole commented:
Has anyone run the e2e conformance tests against this yet?

https://github.com/kubernetes/kubernetes/blob/c6e995a824094a96f7d43a25e897283f83a12997/docs/devel/e2e-tests.md . See the section on conformance tests.

elsonrodriguez committed:
Removing an invalid reference to cadvisor from salt.

It looks like this was left over from https://github.com/kubernetes/kubernetes/pull/23316

elsonrodriguez commented:
@zreigz This was due to a salt error. There was a reference to cadvisor that is no longer valid (it has been removed from the SaltStack).

This PR should fix it: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/9

zreigz commented:
Thanks for fix. I will test it on Monday. BTW now you can also run e2e tests.
```
 KUBERNETES_CONFORMANCE_TEST=y GINKGO_PARALLEL=y go run hack/e2e.go --v --test --test_args="--ginkgo.focus=\[Conformance\] --ginkgo.skip=\[Serial\]"
```

Before I merged your code it was 50/50. I will continue on Monday with your changes

elsonrodriguez committed:
Adding socat package for Redhat systems in Salt

elsonrodriguez commented:
@zreigz @quinton-hoole If run with my pending patch( https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/9 ) I get 74 passed, 18 failed: https://gist.githubusercontent.com/elsonrodriguez/4d221b20c4a0d1e8fbdbf218e9ef390d/raw/a29f0756eabced52ed74e2edc852d443c69fb27b/Openstack%2520Comformance%2520Tests

Two of those tests are most likely related to my proxy setup (Internet and Services test)

For comparison, the AWS provider does 76 passed, 16 failed: https://gist.githubusercontent.com/elsonrodriguez/4d221b20c4a0d1e8fbdbf218e9ef390d/raw/a29f0756eabced52ed74e2edc852d443c69fb27b/AWS%2520Comformance%2520Tests

There's room for improvement, but I think we're looking good for this PR.

Also, @ohtake @batikanu, per https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-213094725 , If we can get all the committers to comment in this thread and provide consent to merge, we can bypass the CLA bot. (P.S., I consent)

elsonrodriguez committed:
Undoing regression of #23311

elsonrodriguez committed:
Triages #7 for OpenStack provider, still needs work.

Fixed the order of fields for basic_auth.

This provider still needs to leverage common.sh for generating proper credentials though.

Also documented a pattern for how to get the SWIFT_SERVER_URL automatically

ohtake committed:
Remove unused code

ohtake committed:
Add braces around environment variable names

Parameter names of Heat str_replace remain unchanged

ohtake committed:
Rename the provider to openstack-heat

There might be several methods to deploy a Kubernetes cluster on OpenStack.
It is better to name the provider more specific.

ohtake commented:
@elsonrodriguez
batikanu and I have already singed the CLA.

* <https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-191648030>
* <https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-213351669>

elsonrodriguez committed:
Changing openstack to openstack-heat in get-kube.sh

elsonrodriguez commented:
@ohtake Everyone has signed the CLA, but per @eparis, explicit consent has to be put in this PR to override the bot.

Something like "I consent to have my commits merged in."

ohtake commented:
Thank you for explanation, @elsonrodriguez

@eparis, I consent to have my commits merged in.

zreigz commented:
@ohtake @elsonrodriguez I had to rebase the branch. There were some conflicts. Please take a look if everything looks good.

Commit:
Replace heat client to openstack to avoid warnings

cheld commented:
In the SIG-OpenStack meet-up there was some discussion and confusion about this PR. So, let me clarify: In Kubernetes project we have some existing OpenStack providers, but they seem half-baked. E.g. dashboard and add-ons are not installed, monitoring not setup, etc...

In this PR we tried to be generic as possible and be as complient with Kubernetes design concept as possible. Some people seem not to like parts of the provining concepts like salt or swift.

I share to some extend these concerns, however if we want to replace them we must do it for all providers - not only OpenStack. Actually, we are already working on improving the provisioning process, but this is not ready yet - hopefully Kubernetes 1.3.

About heat: I assume that those users that have access to it also would like to use it. In addition, it could be leveraged for an node-autoscaling that is similar to GCE. We will also add Monasca monitoring integration in a subsequent PR (also similar to GCE). 

Our overall goal is to give Kubernetes developers and testers a platform that offers a similar functionality as for GCE.

elsonrodriguez commented:
@zreig @quinton-hoole The tests are not passing to the same degree as Amazon anymore (also, the tests are no longer returning consistent results on AWS or OpenStack, these were best of 5):

Openstack: https://gist.githubusercontent.com/elsonrodriguez/71be812acb47a88c85887fbda56c6eb7/raw/6b05d10c44c94e8737cd298831899c10a5fa7586/OpenStack%2520Results
Amazon: https://gist.githubusercontent.com/elsonrodriguez/71be812acb47a88c85887fbda56c6eb7/raw/6b05d10c44c94e8737cd298831899c10a5fa7586/Amazon%2520Results

67 passed for Openstack
72 passed for Amazon

In addition, after some more testing/investigation, here are the current functional issues with this PR:

- OS_AUTH_URL with SSL support: This provider fails on OpenStack installs that use https for the OpenStack api. This was due to some sort of symlink issue with CentOS. There is a pending patch from @francios on Slack to fix this [http://paste.ubuntu.com/16064177/]()
- DNS_SERVER UX issues: On Openstack installs that don't have Neutron with internal DNS, the DNS_SERVER must still be set to a DNS server that can resolve the nodes, otherwise some functionality may not work (`kubectl exec`, `kubectl logs`). Registering nodes with a `--hostname-override <ip address>` doesn't work because of a mismatch between the OpenStack API and the overriden hostname.
- More robust generation of credentials for basic_auth and kubectl based on common.sh

If these issues and test results are not considered fatal to this patch, we should merge in this PR, and tackle remaining problems in separate issues/PRs. Otherwise we're going to run into more rebasing/testing overhead as time goes on, especially since this PR now touches the SaltStack setup and parts of kube-up.

Something else to consider is that contributions to fix these issues have to be in the form of a PR to Fujitsu's PR. Fujitsu has been very responsive, but even so collaborative efforts can be delayed due to timezone differences.

cheld commented:
@zreigz @ohtake can you also take a look

ctrlaltdel committed:
Expose /etc/pki from host to kube-apiserver and controller-manager

CentOS 7 Core nodes running on OpenStack with an SSL-enabled API
endpoint results in the following error without this patch:

F0425 19:00:58.124520       5 server.go:100] Cloud provider could not be initialized: could not init cloud provider "openstack": Post https://my.openstack.cloud:5000/v2.0/tokens: x509: failed to load system roots and no roots provided

The root cause is that the ca-bundle.crt file is actually a symlink
which points to a directory which wasn't previously exposed.

[root@kubernetesstack-master ~]# ls -l /etc/ssl/certs/ca-bundle.crt
lrwxrwxrwx. 1 root root 49 18 nov 11:02 /etc/ssl/certs/ca-bundle.crt -> /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
[root@kubernetesstack-master ~]#

ctrlaltdel commented:
A patch for the OS_AUTH_URL with SSL support issue mentionned by @elsonrodriguez is available here: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/10

elsonrodriguez committed:
Adding workaround for hostname swapping issue in OpenStack kube-up provider.

On reboot, minions hostname would change, causing issue querying Openstack API and registering with kube-apiserver.

This is due to a bug in cloud-init which has been patched but apparently not widely distributed.

Should be harmless on unaffected systems.
https://bugs.launchpad.net/cloud-init/+bug/1246485
http://blog.oddbit.com/2014/12/10/cloudinit-and-the-case-of-the-changing-hostname/

zreigz commented:
There is something wrong after rebase. Provisionig process is not finished correctly for eaxample Docker is not installed

zreigz commented:
OK I know what is wrong. Missing ``` ` ``` in file cluster/saltbase/salt/kube-apiserver/init.sls

Commit:
Add missing single speech mark after rebase

zreigz committed:
Merge pull request #10 from ctrlaltdel/openstack-provider-ssl-fix

Expose /etc/pki from host to kube-apiserver and controller-manager

zreigz committed:
Merge pull request #11 from elsonrodriguez/openstack-provider-hostname

Adding workaround for hostname swapping issue in OpenStack kube-up provider

elsonrodriguez commented:
@zreigz @batikanu @ctrlaltdel Could you all comment something like

"I consent to have my commits merged in to Kubernetes."

Also add a cc to `@quinton-hoole` and `@eparis`.

This is needed for manual override of the CLA bot.

zreigz commented:
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

ctrlaltdel commented:
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

idvoretskyi commented:
@zreigz please, update the documentation https://github.com/kubernetes/kubernetes.github.io/pull/74 with the actual modifications for this provider.

Thank you.

elsonrodriguez commented:
@idvoretskyi Could you clarify the documentation request? What needs to be updated?

xsgordon commented:
@cheld I personally do not strongly object to the use of Heat at this point, but my concern is that the more we rely on projects that are not widely deployed the less useful the provider implementation is. Along similar lines as far as I can tell Monasca isn't represented in OpenStack Foundation user survey responses at all (versus e.g. Ceilometer).

I think we agree that the target is Kubernetes developers and testers, where we disagree is whether those users need or want to set up their own OpenStack cloud to do their work versus being able to run against an existing OpenStack cloud where available. Increased reliance on less widely deployed projects will ensure that they only have one option (deploy their own).

cheld commented:
@xsgordon I would suggest to add _optional_ dependencies even to rare projects like Monasca, LBaaS, ... in case they are required to get fully integrated Kubernetes experience. I agree that this mostly interesting for demo and (integration-)development and might not have that much practical value as of now.  

Heat is not optional and your argument that this provider will not work with all cloud providers is valid. On the other hand, heat is not rare and already supported by the majority. If heat is not acceptable, than we should close this PR, because replacement is not a quick patch. (However, in this case Kubernetes will continue to have no usable provider for OpenStack...)

batikanu commented:
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

xsgordon commented:
@cheld it's not really my argument, I am reflecting the discussion from the SIG (as requested in https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-212288290 ) where the majority of folks present raised the concern that the majority of production OpenStack clouds do not in fact support Heat (only 39%).

My POV is somewhat different than the majority as I work for a vendor whose distribution has been installing Heat OOTB since ~ Havana but none the less I am attempting to accurately convey what was discussed in the context of the SIG (which I should add I do not believe to be the result of confusion about this PR but rather disagreement about what a default provider impl. for OpenStack should look like).

In the context of the SIG we are currently trying to pin down what "ideal state" would look like in terms of optional services, networking, etc., I am attempting to frame a strawman proposal to get things started which you might like to contribute to also as it would like help guide next steps beyond this initial PR?

cheld commented:
So, if I understand you correctly, you suggest to write a design document, first?

elsonrodriguez commented:
@xsgordon , @cheld 

The 39% Heat number comes from this Dataset: https://www.openstack.org/assets/survey/April-2016-User-Survey-Report.pdf

From that document:

- 66% of production Openstack installs have Nova
- 57% have Neutron
- 62% have Glance
- 39% have Heat (as you mentioned)
- 32% have Swift

So Nova, which is the piece that provisions VMs, and what many people consider synonymous with private clouds, is **missing** from 1/3rd of OpenStack installs.

Neutron, which defines networks, and Glance, which handles the upload of OS images, are in a similar state.

Swift, which is analogous to Amazon's S3, and is typically used by kube-up to store k8s artifacts, is down even below Heat.

My point is that in order provide an experience similar to the GCE or AWS kube-up providers, some minimum requirements have to be accepted.

Now outside this scope I think the bar can be set right at ground level, and there is value in that, but the amount of work required to re-implement the basics of a private cloud (or work around the absence of such basics) are out of scope for this PR, and can be addressed in subsequent PRs or other projects (the working doc you mentioned).

xsgordon commented:
@cheld not quite, I think I'm happy to accept the current requirements to get something going but what I am suggesting is further discussion in the context of the SIG on what should be mandatory versus optional going forward as the functionality is expanded.

@elsonrodriguez the percentages are I believe that of the respondents to the question who selected both the project and that they were operating a production environment but yes what I am endeavouring to highlight is that the further down that list you go the less likely the project is to be in an existing environment that someone might want to deploy K8S on.

When we include testing environments the basic infrastructure services receive high percentages overall (90+%) but similarly to the production-only numbers there is a very quick drop off and we see that even Swift (one of the original two projects) is only barely in the majority of respondents clouds overall (53%).

Monasca you will find does not appear at all (unless my eyes deceive me as while I expected a low number I was surprised by this) which I believe means no respondents selected it. According to the report 290 OpenStack users responded to this section of the survey (the survey gives different questions depending on whether you identify as a deployer, app developer, etc. up front).

My concern is simply ensuring that we are making K8S on OpenStack as accessible as possible both for initial deployment but also when making optional features available (for example based on available stats on deployment characteristics I would expect it to make more sense to base scaling logic on Ceilometer than Monasca for example - I believe both the heat-kubernetes and OpenShift-on-OpenStack projects have existing example code for this).

elsonrodriguez commented:
@xsgordon Definitely noted. Past these current requirements we should definitely avoid adding features that require the less adopted OpenStack components. If such features are added, the logic should be cased out for when the components are not available.

@quinton-hoole @a-robinson @eparis We have written consent from all contributors so far, could one of you override the googlebot and set the `cla` label set to `yes`?

Also, if anyone else wants to review this provider and test the functionality that would be most welcome. See here for instructions:

https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes.github.io/blob/openstack-provider-doc/docs/getting-started-guides/openstack-heat.md

You'll have to checkout this branch of kubernetes to test: 
```
git fetch origin pull/21737/head:openstack-provider
git checkout openstack-provider
```

xsgordon commented on cluster/openstack-heat/util.sh:
```diff
@@ -32,7 +32,7 @@ fi
 function verify-prereqs() {
  # Check the OpenStack command-line clients
- for client in swift glance nova heat;
+ for client in swift glance nova openstack;
```
Discussed with @elsonrodriguez on Slack but I believe you still need Heat here or it results in:

openstack: 'stack' is not an openstack command. See 'openstack --help'.
Did you mean one of these?
  backup create
  backup delete
  backup list
  backup restore
  backup show

xsgordon commented on cluster/openstack-heat/util.sh:
```diff
@@ -0,0 +1,274 @@
+# Verify prereqs on host machine
+function verify-prereqs() {
+ # Check the OpenStack command-line clients
+ for client in swift glance nova openstack;
```
Discussed with @elsonrodriguez on Slack but I believe you still need Heat here or it results in:

openstack: 'stack' is not an openstack command. See 'openstack --help'.
Did you mean one of these?
backup create
backup delete
backup list
backup restore
backup show

idvoretskyi replied:
@elsonrodriguez why do we need this specific **python-openstackclient** version?

Are there any restrictions with more recent versions of the client?

elsonrodriguez replied:
@idvoretskyi There's some sort of subcommand difference depending on what version of openstackclient you're using. @xsgordon is the one who ran into the bug, they'll have more details.

No matter what, we are going to have to prescribe a specific version of openstackclient to make sure this works. The individual commands(heat, swift, nova, etc) appear to have more stable arguments and flags. However there were concerns about them being deprecated, and indeed in the most recent versions of the individual commands a huge, distracting deprecation notice was displayed every time a command was issued.

There will always be restrictions for the version of the client, and the restrictions are listed in the documentation, and SHOULD in the future be checked by this provider.

xsgordon replied:
@elsonrodriguez yeah, I plan to create a PR for a check shortly after this lands.

quinton-hoole commented:
LGTM!  Thanks for your patience.  I assume verify-prereqs improvement suggested by @xsgordon can be added in a separate PR.

ctrlaltdel committed:
Automatically detect swift URL if not specified

zreigz committed:
Merge pull request #12 from ctrlaltdel/openstack-provider-swift-url

Automatically detect swift URL

zreigz commented:
This PR and documentation https://github.com/kubernetes/kubernetes.github.io/pull/74 ready for review
PTAL

elsonrodriguez commented:
@quinton-hoole Looks like the bot didn't like you :)

quinton-hoole commented:
@elsonrodriguez That only because commits were added after I gave the LGTM. So bot is being good :-)  New commits also LGTM.  Re-adding label.

quinton-hoole commented:
@elsonrodriguez Oh, I see what you mean, the bot reassigned it.  That was due to a bug in github a week or so ago.  I've manually re-assigned back to myself.

idvoretskyi commented:
@zreigz @elsonrodriguez it is not mandatory, but nice to have:
1) requirements.txt file with the desired `python-'project'client` versions (to install it easily with `pip install -r requirements.txt` command/update the docs with this as well;
2) Replace "OpenStack" with "OpenStack-Heat" on  https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/blob/openstack-provider/cluster/get-kube.sh#L41 as we are expecting other different OpenStack-related providers.

Everything else LGTM. @quinton-hoole, please, take a look, possibly you have some other/extra suggestions.

Commit:
Name adjustment to OpenStack-Heat in cluster/get-kube.sh

zreigz commented:
Thanks for suggestions.  Regarding to:
1) It sounds as a good enhancement for the future. For sure we will do it.
2) Done 
3) **minion** is used as user for instances. It is not only matter of documentation. 
https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/blob/openstack-provider/cluster/openstack-heat/kubernetes-heat/fragments/kube-user.yaml#L4
Finally we get Kubernetes cluster, so is good to keep k8s name convention.

elsonrodriguez commented:
@quinton-hoole Another commit snuck in, could you take another look?

idvoretskyi commented:
@zreigz 
3) please, change that. Don't let people be confused.

elsonrodriguez commented:
@zreigz @idvoretskyi That is a minor cosmetic issue. Almost every other kube-up provider still makes a reference to minions in their usage/code, along with the salt manifests.

I do think the names could be normalized, but that is squarely in the territory of a future PR.

idvoretskyi commented:
@elsonrodriguez @zreigz agree.
LGTM

pwittrock commented:
@zreigz 
@quinton-hoole 

Would you provide an update on the status + ETA for the documentation for this feature as well as add any PRs as they are created?

Not Started / In Progress / In Review / Done

Expected Merge Time:

Thanks
@pwittrock

elsonrodriguez commented:
@pwittrock 
@zreigz 
@quinton-hoole 

Documentation is here: http://kubernetes.io/docs/getting-started-guides/openstack-heat/

And I have a pending PR to fix issues here: https://github.com/kubernetes/kubernetes.github.io/pull/534

Any review/testing would be appreciated.

pwittrock commented:
@elsonrodriguez Is this already available or will it be available in v1.3?  The docs you linked are live, so it may be confusing to users if the feature is not available in the current release.

elsonrodriguez commented:
@pwittrock 1.3 only, I can put a warning regarding the `curl | bash` pattern, and say that only 1.3 releases will work for the tar pattern.
